electrum/gui/qt/wizard/wallet.py

```diff
@@ -0,0 +1,1322 @@
+            self.busy = False
+            self.validate()
+
+        t = threading.Thread(target=unlock_task, args=(client,), daemon=True)
+        t.start()
+
+    def validate(self):
+        if self.password and not self.error:
+            self.apply()
+            self.valid = True
+        else:
+            self.valid = False
+
+    def apply(self):
+        if self.valid:
+            self.wizard_data['password'] = self.password
+
+
+class WCHWXPub(WizardComponent, Logger):
+    def __init__(self, parent, wizard):
+        WizardComponent.__init__(self, parent, wizard, title=_('Retrieving extended public key from hardware'))
+        Logger.__init__(self)
+        self.plugins = wizard.plugins
+        self.plugin = None
+        self._busy = True
+
+        self.xpub = None
+        self.root_fingerprint = None
+        self.label = None
+        self.soft_device_id = None
+
+        self.ok_l = WWLabel(_('Hardware keystore added to wallet'))
+        self.ok_l.setAlignment(Qt.AlignCenter)
+        self.layout().addWidget(self.ok_l)
+
+    def on_ready(self):
+        _name, _info = self.wizard_data['hardware_device']
+        self.plugin = self.plugins.get_plugin(_info.plugin_name)
+        self.title = _('Retrieving extended public key from {} ({})').format(_info.model_name, _info.label)
+
+        device_id = _info.device.id_
+        client = self.plugins.device_manager.client_by_id(device_id, scan_now=False)
+        if not client.handler:
+            client.handler = self.plugin.create_handler(self.wizard)
+
+        cosigner = self.wizard.current_cosigner(self.wizard_data)
+        xtype = cosigner['script_type']
+        derivation = cosigner['derivation_path']
+
+        def get_xpub_task(client, derivation, xtype):
+            try:
+                self.xpub = self.get_xpub_from_client(client, derivation, xtype)
+                self.root_fingerprint = client.request_root_fingerprint_from_device()
+                self.label = client.label()
+                self.soft_device_id = client.get_soft_device_id()
+            except UserFacingException as e:
+                self.error = str(e)
+                self.logger.error(repr(e))
+            except Exception as e:
+                self.error = repr(e)  # TODO: handle user interaction exceptions (e.g. invalid pin) more gracefully
+                self.logger.error(repr(e))
+            self.logger.debug(f'Done retrieve xpub: {self.xpub}')
+            self.busy = False
+            self.validate()
+
+        t = threading.Thread(target=get_xpub_task, args=(client, derivation, xtype), daemon=True)
+        t.start()
+
+    def get_xpub_from_client(self, client, derivation, xtype):  # override for HWW specific client if needed
+        _name, _info = self.wizard_data['hardware_device']
+        if xtype not in self.plugin.SUPPORTED_XTYPES:
+            raise ScriptTypeNotSupported(_('This type of script is not supported with {}').format(_info.model_name))
```

Comment:
You use `DeviceInfo.model_name` in multiple user-facing strings. Note that that field is often None.
Look at `HardwareClientBase.device_model_name`: it is only implemented by trezor, jade, and ledger.
Plugin-agnostic code on master IIRC often uses `device_info.model_name or plugin_name`, see e.g. 
https://github.com/spesmilo/electrum/blob/ddad5f1320993a49327267bf790b485182f13ec0/electrum/base_wizard.py#L355

Maybe the `DeviceInfo` class could have a helper method for this:
`device_info.model_name or device_info.plugin_name or "unknown"`, that always returns a string.

Reply:
I'm in favor of mandating plugins to define this field. In absence, we can initialize it to the plugin name. This approach would let us assume it is always present, avoiding a repeat of `or device_info.plugin_name` at every use of the field.

Reply:
For now I've done the quick fix in ae8c6dfb6347d98d293b213d71d8014b51028cd4, which adds static device names for all HWW. The method can still return None in case a new HWW doesn't define the method, but we now at least get the hint in the IDE that an abstract method is not overridden.

Reply:
added plugin name as default impl for device name

Reply:
Ah I wasn't aware this would/should raise. I'll remove it then, and just rely on a default impl of the method.

Reply:
changed in 8490a2868889e02323c35b8b65c367ebaf0bbc67
I replaced the raise `NotImplementedError`s with `@abstractmethod`  decorators for the methods that must have a subclass definition.